### ISSUE TITLE
Better default VM configuration 

### ISSUE THREAD

**Issue by tsloughter** [gigalixir/gigalixir-getting-started#15]
My understanding is gigalixir runs on Kubernetes, running user's Elixir releases as pods. 

If that is the case and this, https://github.com/gigalixir/gigalixir-getting-started/blob/js/distillery-2.0/rel/vm.args, is the suggested `vm.args` there are a number of changes you should make. I only tried a `find` on the readthedocs and couldn't find any reference to this stuff, sorry if I simply missed it.

Until OTP 23 Erlang is not cgroup aware and will not take into consideration the cgroup limits placed on it when starting up the VM.

The two most important arguments are one for how quickly a scheduler goes to sleep and how many active schedulers are run.

Basically add theses to the `vm.args`:

```
+sbwt none

+S ${SCHEDULERS}
```

Why to set `sbwt` to `none` I have in here https://adoptingerlang.org/docs/production/kubernetes/#container-resources

Where `SCHEDULERS` is an environment variable set for the container in the pod.

I assume you generate a Pod spec for the projects so could just add to the Pod spec:

```
- name: SCHEDULERS
  valueFrom:
    resourceFieldRef:
      containerName: service-discovery
      resource: limits.cpu
      divisor: 1
```

**Comment by christhekeele** [comment]
Based on your description of how to add `SCHEDULERS` to a Pod spec, I assume that `+S ${SCHEDULERS}` will not yet work on gigalixir because it hasn't done that yet?

**Comment by tsloughter** [comment]
@christhekeele as far as I know, yes, that's why I opened this to suggest added it to the PodSpec for everything they generate.

**Comment by jesseshieh** [comment]
Thanks for the suggestion! The reason we don't do this is because we don't actually set a CPU limit in the pod spec. All apps are free to use whatever CPU is available on the kubernetes node it is running on and in 99% of cases, there is plenty of CPU available for all apps since most are memory constrained. We *do* set CPU requests to make sure that in the rare case the kubernetes node is CPU constrained, that each app is allocated their fair share of CPU based on their gigalixir app "size".

**Comment by tsloughter** [comment]
@jesseshieh ah. I'm guessing at this time the k8s nodes aren't massive? So the Erlang nodes aren't booting up with like 32 schedulers? 

With this being the case I think it is even more important to add `+sbwt none`, may even want to require it unless a user can show they need a different value or their app.

I'd still suggest "future proofing" by ldocumenting it for the user in some way, but sounds like for the time being at least users aren't restricted by the cgroup while having many schedulers that are competing for cpu time.

**Comment by jesseshieh** [comment]
It depends what region you're in, but most regions are running with 4-8 cores right now.  Why do you say that it's even more important in this case?

**Comment by tsloughter** [comment]
The `+sbwt none` is even more important because that is even more cpu for pods to burn doing nothing and hurting the performance of themselves (linux CFS scheduler doesn't like it) and especially the other pods on the same k8s node.

**Comment by jesseshieh** [comment]
Oh I see what you mean now. Thanks for explaining. I don't think the busy wait actually uses up 100% of the CPU available to it and all the kubernetes nodes I'm looking at are well below 100% CPU which suggests to me that allowing all the BEAMs to busy wait isn't causing a problem so far.

Also, assuming the BEAM has enough CPU to busy wait, I think disabling it might adversely affect performance. I could definitely be wrong though so let me know if I'm misunderstanding something.

**Comment by tsloughter** [comment]
Because you have no limits set the fact that `sbwt` is not `none` likely only results in wasted cpu cycles and higher overall usage, and not poor performance because it won't be throttling pods because of the busy wait. It is usually a killer in shared environments, so why I brought it up.

I'd suggest taking a look at what difference it makes. There may not be any major adverse effect currently if you are very over provisioned but you should still see a nice drop in CPU usage. And there are multiple values, like `very_short` that could be used instead if it turns out an app actually has a usage pattern that benefits from the busy wait.

I can go ahead and close this issue shortly, but it still might be useful to include some of this in your docs and I hope it helps when you do find yourself in the happy position of needing to limit usage and have nodes hitting their capacity :)

**Comment by jesseshieh** [comment]
Ah that all makes a lot of sense. Thanks for all the help!
